### ISSUE TITLE
php shell / cli script and namespaces

### ISSUE THREAD

**Issue by Nickwiz** [dense-analysis/ale#4235]
### Discussed in https://github.com/dense-analysis/ale/discussions/4232

<div type='discussions-op-text'>

<sup>Originally posted by **Nickwiz** June 14, 2022</sup>
When using ALE with PHP as cgi, shell script ... and using `namespace` the first lines can be something like:

```php
#! /usr/bin/env php
<?php
namespace CLI;
echo "OK\n";
```

This produces ALE error:

> Namespace declaration statement has to be the very first statement or after any declare call in the script

It complains about the *shebang* and note that PHP does so if the file is redirected instead of used as an argument, i.e:

**OK:**

```sh
$ php -l -d error_reporting=E_ALL -d display_errors=1 -d log_errors=0 test.php 
No syntax errors detected in test.php

$ php test.php
OK

$ ./test.php
OK
```

**BAD:**
```sh
$ php -l -d error_reporting=E_ALL -d display_errors=1 -d log_errors=0 <test.php
Fatal error: Namespace declaration statement has to be the very first statement or
after any declare call in the script in Standard input code on line 3
Errors parsing Standard input code

$ php< test.php
PHP Fatal error:  Namespace declaration statement has to be the very first statement or 
after any declare call in the script in Standard input code on line 3
```

Is there an option to rectify for this somewhere? If not, it should perhaps be reported as a bug, but not sure. 

I do not know the ALE code and not sure why it redirects the file instead of using it as an argument.

Have not found any options in the *Command line options* section that could help in this regard. 
https://www.php.net/manual/en/features.commandline.options.php</div>

**Comment by hsanson** [comment]
You can see the *command* that ALE uses at [php.vim#L37](https://github.com/dense-analysis/ale/blob/master/ale_linters/php/php.vim#L37)

```
call ale#linter#Define('php', {
\   'name': 'php',
\   'executable': {b -> ale#Var(b, 'php_php_executable')},
\   'output_stream': 'stdout',
\   'command': '%e -l -d error_reporting=E_ALL -d display_errors=1 -d log_errors=0 --',
\   'callback': 'ale_linters#php#php#Handle',
\})
```

By default ALE uses stdin (redirection) for passing the data to the linters but this can be changed by setting the lint_file option to "1".  See `:h ale#linter#Define`.  Unfortunatelly this change would require changes to ALEs code as this cannot be changed via config and also reading files from disk disables linting while typing. Read the docs in lint_file option for more details.

**Comment by rymdbar** [comment]
I believe this is a duplicate of #1445. PHP 7.4.0 and later includes a fix for the issue.

PHP 7.4 was released in 2019, and in in 2021 PHP 7.3 was end-of-lifed released according to [php.net][eol]. Also, I can not reproduce (with a modern version).

Does this still happen @Nickwiz or can the issue be closed?

[php]: https://www.php.net/supported-versions.php
[eol]: https://www.php.net/eol.php

**Comment by Nickwiz** [comment]
> Does this still happen [@Nickwiz](https://github.com/Nickwiz) or can the issue be closed?

No. Can not reproduce with v8. Looking at old disks it looks like this was under 7.2 but can not be sure. 
(OP should have specified that lazy bastard ;-P )

This one if skimmed logs are interpreted correctly:

https://launchpad.net/ubuntu/+source/php7.2/7.2.24-0ubuntu0.18.04.12

Last 7.2 security update was February 2023:

https://launchpad.net/ubuntu/+source/php7.2/7.2.24-0ubuntu0.18.04.17

Long since moved away from Ubuntu (and 7.2 PHP). IIRC it was on a project requiring 7-series back then.
